### DISCUSSION TITLE
setError: Type is not assignable to type 'Path<TFormValues>'

### DISCUSSION THREAD

**Discussion by elidrissidev** [react-hook-form/react-hook-form#8192] [Q&A] [answered]
I have been battling with this for a few hours now, I have a generic Form component that takes care of calling `useForm` and setting `yup` resolver, but I also want to include server-side error handling.

I tried the approach below but I'm getting a Typescript error on `setError`'s first parameter: `Type 'string & keyof TFormValues' is not assignable to type 'Path<TFormValues>'`. I already saw a alt approach in [React Hook Form Server-side Validation article](https://www.carlrippon.com/react-hook-form-server-validation/), but I'd like to know if there's another way without passing `setError` through and changing its type. Thanks in advance 🙏.

```tsx
// Form.tsx
import { useForm, UseFormReturn, SubmitHandler } from 'react-hook-form'

import { isValidationError } from '@/utils/api/validation'

type FormProps<TFormValues> = {
  children: (methods: UseFormReturn<TFormValues>) => React.ReactNode
  onSubmit: SubmitHandler<TFormValues>
}

export function Form<TFormValues>({ children, onSubmit }: FormProps<TFormValues>) {
  const methods = useForm<TFormValues>()

  const handleSubmit: React.FormEventHandler = e =>
    methods
      .handleSubmit(onSubmit)(e)
      .catch(err => {
        if (isValidationError<TFormValues>(err)) {
          const validationErrors = err.response!.data.errors
          for (let fieldName in validationErrors) {
            methods.setError(fieldName, { message: validationErrors[fieldName][0] })
          }
        }
      })

  return <form onSubmit={handleSubmit}>{children(methods)}</form>
}
```

```ts
// utils/api/validation.ts
import axios, { AxiosError } from 'axios'

interface ValidationErrorResponse<TFields> {
  errors: {
    [field in keyof TFields]: string[]
  }
}

export function isValidationError<TFields>(
  error: any
): error is AxiosError<ValidationErrorResponse<TFields>> {
  return axios.isAxiosError(error) && 'errors' in error.response?.data
}
```

**Comment by bluebill1049** [comment]
will this feature request resolve your problem above? https://github.com/react-hook-form/react-hook-form/issues/7661

**Reply by elidrissidev** [reply]
Hey, thanks for replying.

My issue is not really with global errors, but rather just making Typescript satisfied about  my `setError` usage, the errors I'm trying to handle are all part of form fields. I only seem to encounter this issue in this generic Form component as I've used it before in another component without any issue.

**Comment by leapful** [comment, marked as answer]
@elidrissidev : 

It seems your type definition for ValidationErrorResponse is not matched with the expected react-hook-form type and then missing of type assigning when looping with "for ... in ..." statement. 

Here is the update version:

```
interface ValidationErrorResponse<TFields> {
  errors: Record<FieldPath<TFields>, string[]>
}

const hasOwnProperty = <T extends object> (data: T, key: any): key is keyof T => {
  return Object.prototype.hasOwnProperty.call(data, key);
}

// then use custom "hasOwnProperty" function to assign the type.

for (const fieldName in validationErrors) {
   if (hasOwnProperty(validationErrors, fieldName)) {
       methods.setError(fieldName, {message: validationErrors[fieldName][0]});
   }
}
```

NOTE: Please always remember to use "Object.prototype.hasOwnProperty" when using statement "for ... in ..." to avoid potential issue on prototype properties, it's a best practice in javascript.

Referrer: https://eslint.org/docs/rules/guard-for-in

**Reply by elidrissidev** [reply]
Thank you so much! Will try that tomorrow and report back.

**Reply by elidrissidev** [reply]
This works perfectly. Thank you!

**Reply by bluebill1049** [reply]
thanks @leapful! it's been a while since I have seen you again here. ❤️  hope everything is doing well. :)

**Comment by kiranmarshall** [comment]
I have an alternative solution which fits my use case with little fuss:

```
const { control, formState, register, getValues, setError, clearErrors } = useFormContext<
      Interface & { optionalValue?: boolean }
   >();
```

This is given the following constraints:
- The error occurs from a smaller part the form, hence using `useFormContext()`. I did not try this at the top level using `useForm`
- The error is just reacting to a computed property resulting in a boolean
- The form cannot be submitted with this error, so we can guarantee that it doesn't need to be part of the top level interface that react hook form validates with. 

Do leave a comment if you can see any potential issues with this implementation.
